### ISSUE TITLE
Q: nfs-ganesha uses a lot of wall time. If a clock jump occurs, does it affect functionality?

### ISSUE THREAD

**Issue by Haroldll** [nfs-ganesha/nfs-ganesha#927]
nfs-ganesha uses a lot of wall time. For example, time(), gettimeofday().
If the server clock is changed, for example, the clock module is broken, the time becomes the original time.
Some features of ganesha may be affected when the clock is modified, is this a potential problem？
Would it be better to use CPU time?

just like this:
```
static unsigned int _valid_lease(nfs_client_id_t *clientid)
{
	time_t t;

	if (clientid->cid_confirmed == EXPIRED_CLIENT_ID)
		return 0;

	if (clientid->cid_lease_reservations != 0)
		return nfs_param.nfsv4_param.lease_lifetime;

	t = time(NULL);

	if (clientid->cid_last_renew + nfs_param.nfsv4_param.lease_lifetime > t)
		return (clientid->cid_last_renew +
			nfs_param.nfsv4_param.lease_lifetime) - t;

	return 0;
}
```
If it becomes a time in the past, the lease may not time out, even if no OP is received from the client.

**Comment by dang** [comment]
No, the proper solution would be to use CLOCK_MONOTONIC for cases that cannot go backwards.  However, the *real* proper solution is to run NTP on all servers, and to start that before you start ganesha.  That way there won't be a problem.

**Comment by ffilz** [comment]
Bumping things that got depressed to to tagging other issues.

**Comment by Haroldll** [comment]
> No, the proper solution would be to use CLOCK_MONOTONIC for cases that cannot go backwards. However, the _real_ proper solution is to run NTP on all servers, and to start that before you start ganesha. That way there won't be a problem.

Thank you for your reply. Yes, NTP is the best solution. Under this premise, is it possible that the clock is modified incorrectly, changed. Of course, it will be resynchronized, but there is a process of change.
The process of clock change affects functions that depend on the wall time.

**Comment by dang** [comment]
There are lots of things an administrator could do to a server that could temporarily or permanently mess with Ganesha's operation.  I agree that changing the time is one of them, but I'm not sure it's worth trying to anticipate all the possible issues and guard against them.  I suspect the most pragmatic response is "Don't do that".

If someone is interested in starting a project to audit all the time uses in Ganesha and fix the ones that cannot go backwards, I'm sure we'd be interested in taking the patches.

**Comment by Haroldll** [comment]
ok, I will try

**Comment by Haroldll** [comment]
https://review.gerrithub.io/c/ffilz/nfs-ganesha/+/555292
@dang Please review

**Comment by ffilz** [comment]
I don't think the patch is a good way to solve the issue. Is this still a problem and if so, can we find another solution?
